apps/btc_family/btc_txn.c

```diff
@@ -247,31 +286,183 @@ static bool fetch_transaction_meta(btc_query_t *query) {
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+  }
+  if (true == zero_value_transaction) {
+    // do not allow zero valued transaction
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    return false;
+  }
+  return true;
+}
+
+static bool get_user_verification() {
+  char title[20] = "";
+  char value[100] = "";
+  char address[100] = "";
+
+  for (int idx = 0; idx < btc_txn_context->metadata.output_count; idx++) {
+    btc_sign_txn_output_t *output = &btc_txn_context->outputs[idx];
+    btc_sign_txn_output_script_pub_key_t *script = &output->script_pub_key;
+    snprintf(title, sizeof(title), UI_TEXT_BTC_RECEIVER, (idx + 1));
+
+    if (true == output->is_change) {
+      // do not show the change output UTXOs
+      continue;
+    }
+    format_value(output->value, value, sizeof(value));
+    script_output_to_address(
+        script->bytes, script->size, address, sizeof(address));
+    if (!core_scroll_page(title, address, btc_send_error) ||
+        !core_scroll_page(title, value, btc_send_error)) {
+      return false;
+    }
+  }
+
+  // calculate fee in various forms
+  uint64_t max_fee = get_transaction_fee_threshold(btc_txn_context);
+  uint64_t fee_in_satoshi = 0;
+
+  if (!btc_get_txn_fee(btc_txn_context, &fee_in_satoshi)) {
+    // The transaction is overspending
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    return false;
+  }
+
+  // all the receivers are verified, check fee limit & show the fee
+  // validate fee limit is not too high and acceptable to user
+  if (fee_in_satoshi > max_fee &&
+      !core_confirmation(ui_text_warning_txn_fee_too_high, btc_send_error)) {
+    return false;
+  }
+
+  format_value(fee_in_satoshi, value, sizeof(value));
+  if (!core_scroll_page(UI_TEXT_BTC_FEE, value, btc_send_error)) {
+    return false;
+  }
+  return true;
+}
+
+static bool sign_input_utxo(btc_query_t *query) {
+  if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) ||
+      !check_which_request(query, BTC_SIGN_TXN_REQUEST_SIGNATURE_TAG)) {
+    return false;
+  }
+
+  // TODO: Sign each input and send the signature
+  return false;
+}
+
 /*****************************************************************************
  * GLOBAL FUNCTIONS
  *****************************************************************************/
 
 void btc_sign_transaction(btc_query_t *query) {
-  btc_txn_context = (txn_context *)malloc(sizeof(txn_context));
-  memzero(btc_txn_context, sizeof(txn_context));
+  btc_txn_context = (btc_txn_context_t *)malloc(sizeof(btc_txn_context_t));
+  memzero(btc_txn_context, sizeof(btc_txn_context_t));
 
-  if (!handle_initiate_query(query) && !fetch_transaction_meta(query)) {
+  if (!handle_initiate_query(query) && !fetch_transaction_meta(query) &&
+      !fetch_input_utxo(query) && !fetch_output_utxo(query) &&
```

Comment:
Could rename these functions as `fetch_valid_input_utxos` and `fetch_txn_outputs` and in comments add validation meaning.

Reply:
changed